v2/api/containerservice/customizations/managed_cluster_extensions.go

```diff
@@ -120,3 +123,61 @@ func secretsToWrite(obj *containerservice.ManagedCluster, adminCreds string, use
 
 	return collector.Values()
 }
+
+var _ extensions.PreReconciliationChecker = &ManagedClusterExtension{}
+
+// If a managed cluster has a provisioningState in this set, it will reject any attempt to PUT a new state out of hand;
+// so there's no point in even trying. This is true even if the PUT we're doing will have no effect on the state of the
+// cluster.
+var blockingManagedClusterProvisioningStates = set.Make(
+	"Creating",
+	"Updating",
+	"Scaling",
+	"Deleting",
+	"Migrating",
+	"Upgrading",
+	"Stopping",
+	"Starting",
+	"RotatingClusterCertificates",
+	"ReconcilingClusterCertificates",
+	"RotatingClusterStaticTokens",
+	"ReconcilingClusterETCDCertificates",
+	"RotatingServiceAccountSigningKeysInternal",
+	"RotatingServiceAccountSigningKeysExternal",
+	"Canceling",
+)
+
+func (ext *ManagedClusterExtension) PreReconcileCheck(
+	_ context.Context,
+	obj genruntime.MetaObject,
+	_ kubeclient.Client,
+	_ *genericarmclient.GenericClient,
+	_ logr.Logger,
+	_ extensions.PreReconcileCheckFunc,
+) (extensions.PreReconcileCheckResult, error) {
+	// This has to be the current hub storage version. It will need to be updated
+	// if the hub storage version changes.
+	managedCluster, ok := obj.(*containerservice.ManagedCluster)
+	if !ok {
+		return extensions.SkipReconcile("Expected Managed Cluster"),
+			errors.Errorf("cannot run on unknown resource type %T, expected *containerservice.ManagedCluster", obj)
+	}
+
+	// Type assert that we are the hub type. This will fail to compile if
+	// the hub type has been changed but this extension has not
+	var _ conversion.Hub = managedCluster
+
+	// If the cluster is in a state that will reject any PUT, then we should skip reconciliation
+	// as there's no point in even trying.
+	// This allows us to "play nice with others" and not use up request quota attempting to make changes when we
+	// already know those attempts will fail.
+	if provisioningState := managedCluster.Status.ProvisioningState; provisioningState != nil {
+		if blockingManagedClusterProvisioningStates.Contains(*provisioningState) {
+			return extensions.SkipReconcile(
+					fmt.Sprintf("Managed cluster is in provisioning state %q", *provisioningState)),
```

Comment:
Some details about the resource would be helpful here?

Reply:
We don't need them here; the message is consumed twice - once via a logger that already includes the details of the resource itself, and once as the condition on the resource in the cluster. No ambiguity on what is being described.